File: Coding-Patterns/Coding-Patterns-CSharp/CodingPatterns/04_dynamic_programming/LongestIncreasingSubsequence.cs
// LeetCode 300 - Longest Increasing Subsequence
// Difficulty: Medium
// Pattern: Patience sort with binary search
//
// Time: O(n log n)  Space: O(n)

namespace CodingPatterns.DynamicProgramming;

public class LongestIncreasingSubsequence
{
    public int LengthOfLIS(int[] nums)
    {
        var tails = new List<int>();
        foreach (var num in nums)
        {
            int lo = 0, hi = tails.Count;
            while (lo < hi)
            {
                int mid = (lo + hi) / 2;
                if (tails[mid] < num) lo = mid + 1;
                else hi = mid;
            }

            int pos = lo;
            if (pos == tails.Count) tails.Add(num);
            else tails[pos] = num;
        }
        return tails.Count;
    }

    // ---- Tests ----
    public static void Run()
    {
        var sol = new LongestIncreasingSubsequence();

        Console.WriteLine(sol.LengthOfLIS(new[] { 10, 9, 2, 5, 3, 7, 101, 18 }));  // 4
        Console.WriteLine(sol.LengthOfLIS(new[] { 0, 1, 0, 3, 2, 3 }));            // 4
        Console.WriteLine(sol.LengthOfLIS(new[] { 7, 7, 7, 7 }));                  // 1
        Console.WriteLine(sol.LengthOfLIS(Array.Empty<int>()));                    // 0
    }
}

File: Coding-Patterns/Coding-Patterns-Python/04_dynamic_programming/longest_increasing_subsequence.py
# LeetCode 300 - Longest Increasing Subsequence
# Difficulty: Medium
# Pattern: 1D DP (or patience sort with binary search for O(n log n))
#
# Problem: Given an integer array nums, return the length of the
# longest strictly increasing subsequence.
#
# Approach 1 (O(n^2) DP): dp[i] = LIS ending at index i.
# Approach 2 (O(n log n)): Maintain a "tails" array where tails[i]
# is the smallest tail element of all increasing subsequences of length i+1.
# Use binary search to update tails.
#
# Time: O(n log n)  Space: O(n)

from typing import List
import bisect

def length_of_lis(nums: List[int]) -> int:
    tails = []  # tails[i] = smallest tail for LIS of length i+1
    for num in nums:
        pos = bisect.bisect_left(tails, num)
        if pos == len(tails):
            tails.append(num)
        else:
            tails[pos] = num  # replace with smaller tail
    return len(tails)


# ---- Tests ----
if __name__ == "__main__":
    assert length_of_lis([10, 9, 2, 5, 3, 7, 101, 18]) == 4  # [2,3,7,101]
    assert length_of_lis([0, 1, 0, 3, 2, 3]) == 4
    assert length_of_lis([7, 7, 7, 7]) == 1
    print("All tests passed.")
